친구 / 차단 관련 API

seongheonsim opened the pull request:
@holeman79 
안녕하세요, 원진님. 친구 및 차단 관련 기능 API 개발 완료해서 PR 올립니다.
개발된 내용은 다음과 같습니다.

- Friend
  - 친구 요청 API
  - 친구 요청 수락 API
    - 거절 시, 해당 데이터 삭제
  - 친구 목록 조회 API (+ 닉네임 검색)
- BlockedMember
  - 회원 차단 API
    - 회원 차단 시, 해당 회원에 대한 친구 데이터가 존재한다면 해당 데이터 삭제
  - 회원 차단 해제 API
  - 차단 목록 조회 API (+ 닉네임 검색)

천천히 확인 부탁드립니다!

Commit:
feat(Friendship) : 친구 목록 및 친구 닉네임 검색 API 추가

Commit:
feat(Friendship) : 친구 목록 및 친구 닉네임 검색 API 문서 추가

Commit:
feat(Friendship) : 친구 신청 및 요청 처리 API 추가

Commit:
feat(Friendship) : 친구 조회 API 페이징 처리 추가

Commit:
feat(Friendship) : 친구 조회 API 쿼리 수정

Commit:
feat(Friendship) : 친구 조회 API 엔드포인트 수정

Commit:
feat(Friendship) : 친구 요청 / 승낙 / 거부 API 문서 추가

Commit:
Merge branch 'dev' into feat/member_friends

Commit:
feat(Friend / BlockedMember) : 친구 도메인 관련 파일 이동 및 차단 관련 기능 API 추가

1. Friendship -> Friend 도메인명 변경
2. Friend 관련 파일 member 디렉토리로 이동
3. 차단 관련 기능 API 추가

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/application/MemberFriendAddService.kt:
```diff
@@ -0,0 +1,36 @@
+    fun add(inviterId: Long, inviteeId: Long) {
+        // 차단 여부 조회
+        val isBlocked = blockedMemberExistDao.existByRequesterOrTargetMember(inviterId, inviteeId)
+        if (isBlocked) throw FriendInputException("친구 신청이 불가능한 유저입니다.")
```
BlockedMemberExistDao와 FriendExistDao를 합쳐서 FriendValidateDao로 만들고 그 안에서 validation 로직과 Exception을 던지는게 어떨까요?

seongheonsim replied:
@holeman79 
해당 부분 반영 완료했습니다!

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/api/MemberFriendAddApi.kt:
```diff
@@ -0,0 +1,27 @@
+class MemberFriendAddApi(
+    private val memberFriendAddService: MemberFriendAddService
+) {
+    @PostMapping("/{inviterId}/friends")
```
이 url이 나타내는 의미는 member 밑에 존재하는 friend를 추가하겠다는 것을 나타내고 있지만 실제로는 member 도메인 클래스 안에는 friend를 갖고 있지 않습니다. friend 도메인 패키지를 따로 만들어서 /api/friends @PostMapping 이렇게 처리하는 것에 대해 성헌님은 어떻게 생각하시나요?

seongheonsim replied:
@holeman79 
해당 부분 반영 완료 했으며, 저번 모각코에서 나눴던 내용처럼 BlockedMember 쪽도 동일하게 분리했습니다!

holeman79 commented on src/test/kotlin/com/beside/groubing/groubingserver/domain/member/api/MemberFriendFindApiTest.kt:
```diff
@@ -0,0 +1,112 @@
+            }
+        }
+    }
+})
```
친구나 차단에 대한 API 문서 Test 코드만 있는거 같은데 기능에 대한 테스트 코드는 따로 작성은 안하셨나요? 테스트는 어떻게 해보셨을까요?

seongheonsim replied:
@holeman79 
급한 마음에 테스트 코드는 작성하지 못하고 포스트맨으로 테스트 진행했습니다ㅠㅠ

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/global/response/PageResponse.kt:
```diff
@@ -3,7 +3,7 @@ package com.beside.groubing.groubingserver.global.response
+    val contents: List<T> = page.content
 
     val totalElements: Long = page.totalElements
 
```
페이징에 대한 객체를 추가한 부분 너무 좋습니다. 근데 총 갯수만 있는 거 같은데 현재 페이지가 몇 번째 페이지인지, 총 페이지 수, 현재 Element 개수 등도 함께 표시하면 어떨까요? 아래는 제가 Java 버전으로 사용하던 PageDto인데 참조하시면 좋을 듯 합니다ㅎㅎ
```
public class PageDto<T> {
    private List<T> contents;

    private long totalElements;

    private int numberOfElements;

    private int currentPage;

    private int totalPages;

    private int size;

    private boolean first;

    private boolean last;

    public PageDto(Page<T> page) {
        this.contents = page.getContent();
        this.totalElements = page.getTotalElements();
        this.numberOfElements = page.getNumberOfElements();
        this.totalPages = page.getTotalPages();
        this.currentPage = page.getNumber() + 1;
        this.size = page.getSize();
        this.first = page.isFirst();
        this.last = page.isLast();
    }

    public static <T> PageDto<T> of(Page<T> page) {
        return new PageDto<>(page);
    }

    public static <T> PageDto<T> of(List<T> contents, Pageable pageable, long totalCount) {
        return of(new PageImpl<>(contents, pageable, totalCount));
    }
}
```

seongheonsim replied:
@holeman79 
해당 클래스는 원진님께서 만들어주신 클래스라 예시로 올려주신 코드가 해당 클래스에 적용되어 있습니다~!

Commit:
fix(Friend, BlockedMember) : 친구 및 차단 회원 도메인 분리

Commit:
fix(Friend) : 친구 신청 검증 로직 변경

Commit:
fix(Core) : Kotlin Version 변경 / asciidoctor 경고 메세지 수정

seongheonsim commented:
@holeman79 
프로젝트 빌드 시, aciidoctor 쪽에서 경고 메세지가 뜨길래 이를 해결하기 위한 코드를 추가했습니다.
추가적으로 Kotlin 및 Gradle 버전을 최신 버전으로 올렸습니다.

<img width="1582" alt="Screenshot 2023-07-20 at 22 59 33" src="https://github.com/beside-14/groubing-server/assets/60208293/4c32ed1b-b0cb-48eb-9b57-cdc57a19d278">

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/blocked/api/BlockedMemberFindApi.kt:
```diff
@@ -0,0 +1,32 @@
+    ): ApiResponse<PageResponse<BlockedMemberResponse>> {
+        val response = when (nickname.isNullOrBlank()) {
+            true -> blockedMemberFindService.findById(requesterId, pageable)
+            false -> blockedMemberFindService.findByIdAndNickname(requesterId, pageable, nickname)
```
닉네임 찾기는 백엔드로 보내는 쿼리가 실행되는게 아닌 듯 합니다. (카카오톡에서 이름으로 검색처럼 프론트에서 하는 것으로 보임). 그리고 친구 목록이랑 차단 친구 목록은 와이어프레임상 페이징은 없는 것으로 보입니다!

seongheonsim replied:
@holeman79 
앗 제가 제대로 이해한게 맞는지 모르겠는데, 원진님 말씀은 프론트에서 친구 및 차단 회원 목록 조회 API 호출 시, 모든 데이터를 리턴하고 해당 데이터를 프론트에서 가지고 있어 이를 기준으로 프론트에서 닉네임 검색을 수행한다는 말씀이실까요?

만약 제가 이해한게 맞다면 원진님 말씀처럼 친구 및 차단 회원 목록 조회 API 를 전체 조회로 변경해야 할 것 같은데 이 부분은 프론트 분들과 한 번 이야기하는게 좋을 것 같다는 생각이 들어요! 원진님 의견은 어떠신가요?!

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/blocked/api/BlockedMemberFindApi.kt:
```diff
@@ -0,0 +1,32 @@
+) {
+    @GetMapping
+    fun find(
+        @AuthenticationPrincipal requesterId: Long,
```
@AuthenticationPrincipal 은 이전에 member 도메인에서 사용했었던 memberId로 통일하는게 어떨까요?

seongheonsim replied:
@holeman79 
해당 피드백 Friend / BlockedMember 모두 반영 완료 했습니다! 🫡

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/blocked/api/BlockingMemberApi.kt:
```diff
@@ -0,0 +1,23 @@
+        @AuthenticationPrincipal requesterId: Long,
+        @RequestBody request: BlockingMemberRequest
+    ) {
+        blockingMemberService.blocking(requesterId, request.targetMemberId)
```
메소드 명은 blocking보다 동사형으로 block을 사용하는게 어떨까요?

seongheonsim replied:
@holeman79 
해당 피드백 반영 완료 했습니다~!🫡

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/blocked/application/BlockingMemberService.kt:
```diff
@@ -0,0 +1,38 @@
+    fun blocking(requesterId: Long, targetMemberId: Long) {
+        // 차단 여부 확인
+        val isBlocked = blockedMemberExistDao.existByRequesterOrTargetMember(requesterId, targetMemberId)
+        if (isBlocked) throw BlockedMemberInputException("이미 차단 되었거나 차단이 불가능한 유저입니다.")
```
BlockedMemberInputException 던지는 것도 blockedMemberExistDao 내에서 처리하는게 어떨까요?

seongheonsim replied:
@holeman79 
해당 피드백 반영 완료 했습니다~!🫡
다만 `BlockedMemberExistDao` 클래스의 경우 `Friend` 도메인에서도 사용하고 있어 예외 메세지를 두 도메인에서 바로 납득할 수 있는 내용으로 변경 했습니다.

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/blocked/application/BlockingMemberService.kt:
```diff
@@ -0,0 +1,38 @@
+        if (isBlocked) throw BlockedMemberInputException("이미 차단 되었거나 차단이 불가능한 유저입니다.")
+
+        // 친구 여부 확인 > 친구인 경우 친구 관계 삭제
+        val friendships = friendFindDao.findByInviterOrInvitee(requesterId, targetMemberId)
```
findByInviterOrInvitee 명칭보다 findFriendShips가 어떨까요?

seongheonsim replied:
@holeman79 
제안주신 메소드명이 훨씬 더 직관적이네요! 해당 피드백 반영 완료 했습니다~!

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/blocked/dao/BlockedMemberExistDao.kt:
```diff
@@ -0,0 +1,23 @@
+            .from(blockedMember)
+            .where(
+                blockedMember.requester.id.eq(requesterId).and(blockedMember.targetMember.id.eq(targetMemberId)).or(
+                    blockedMember.requester.id.eq(targetMemberId).and(blockedMember.targetMember.id.eq(requesterId))
```
상대방이 나를 차단했으면 나는 상대방을 차단할 수 없나요?

seongheonsim replied:
@holeman79 
제가 생각 했을 땐 상대가 나를 차단한 경우, 상대와 관련된 정보(유저 정보, 빙고 등)가 내 피드, 검색 등에 노출이 되지 않아야 한다고 생각해서 이렇게 코드를 작성 했습니다..! 그래서 이 부분은 기획팀분들께 한 번 여쭤보고 정책을 정해야 할 것 같습니다..!

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/blocked/dao/BlockedMemberExistDao.kt:
```diff
@@ -0,0 +1,23 @@
+                    blockedMember.requester.id.eq(targetMemberId).and(blockedMember.targetMember.id.eq(requesterId))
+                )
+            )
+            .fetchOne()
```
https://jojoldu.tistory.com/516
이동욱님 블로그를 보면 exist 쿼리를 사용할 때 count()로 집계해서 > 0 로직을 사용하는 것보다 fetchFirst()를 사용하는 것이 쿼리 성능이 더 좋다고 합니다. count() 쿼리는 전체 데이터를 조회하지만 fetchFirst의 경우 일치하는 데이터가 나올 경우 바로 종료하기 때문입니다! 참고하시면 좋을 듯 합니다ㅎㅎ

```
queryFactory
                .selectOne()
                .from(threatActor)
                .where(threatActor.aliases.any().in(names))
                .fetchFirst() != null;
```

seongheonsim replied:
@holeman79 
엇 말씀해주신 내용은 모르고 있었는데 알려주셔서 감사합니다!
해당 피드백 반영 완료 했습니다!

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/friend/application/FriendDeleteService.kt:
```diff
@@ -0,0 +1,15 @@
+    private val friendRepository: FriendRepository
+) {
+    fun delete(id: Long) {
+        friendRepository.deleteById(id)
```
아예 삭제를 해버리는게 아니라 거절 상태가 있어야 나중에 이력을 보여주거나 데이터 활용 측면에서 더 좋지 않을까요?

seongheonsim replied:
@holeman79 
이력이라고 하신 부분은 제가 차단 처리에 대한 내역을 조회 하는 것을 의미하는 것 일까요?!
이 부분은 모각코에서 조금 더 이야기 나눠보시는건 어떠세요?!

holeman79 commented on src/main/kotlin/com/beside/groubing/groubingserver/domain/member/api/SignUpApi.kt:
```diff
@@ -20,7 +20,7 @@ class SignUpApi(
         @Validated
         request: SignUpRequest
-    ): ApiResponse<MemberResponse> {
+    ): ApiResponse<MemberLoginResponse> {
```
이거 왜 MemberResponse 에서 다시 MemberLoginResponse 로 돌아갔나요? 예전에 MemberResponse로 통일 했던 걸로 아는데 갑자기 이전 코드로 다시 바뀌었네요?

seongheonsim replied:
@holeman79 
처음 Member 도메인 내에 Friend / BlockedMember 도메인을 포함해 개발하면서 회원 상세 응답 모델 클래스를 만들게 되었는데, MemberResponse 와 MemberDetailResponse 가 서로 비슷한 의미를 가질 수 있겠다고 생각해서 수정 했던 것 같습니다..!

Friend / BlockedMember 도메인을 별도로 빼는 과정 중에 다시 원복 했어야 했는데 말씀해주신 부분을 놓친 것 같습니다ㅠㅠ
해당 피드백 다시 원복 완료 했습니다!

Commit:
fix(Friend, BlockedMember) : @AuthenticationPrincipal 변수명 통일

Commit:
fix(BlockedMember) : 클래스명 및 메소드명 동사형으로 수정

Commit:
fix(BlockedMember) : 차단 여부 체크에 대한 예외 처리 위치 수정

Commit:
fix(BlockedMember) : findByInviterOrInvitee 메소드명 수정

Commit:
fix(Member) : MemberLoginResponse -> MemberResponse 클래스명 변경

Commit:
fix(BlockedMember) : 차단 여부 조회 시 fetchFirst() 로 처리하도록 수정

Commit:
Merge branch 'dev' into feat/member_friends

Commit:
fix(BlockedMember) :

- 차단 여부 체크 시, '나' 를 기준으로만 조회하도록 수정

Commit:
fix(Friend) :

- 친구 요청 시, 신규 친구 요청인 경우 새로운 데이터를 생성하고 거절된 데이터가 존재하는 경우 해당 데이터를 PENDING 상태로 변경하여 업데이트 하도록 수정
- 친구 요청 거절 시, 상태를 REJECT 로 변경하여 업데이트 하도록 수정

Commit:
fix(Friend) : 친구 및 차단 회원 목록 조회 로직 수정

- 친구 및 차단 회원 닉네임 검색 API 삭제
- 친구 및 차단 회원 조회 시, 전체 데이터 조회하도록 수정

Commit:
Merge branch 'dev' into feat/member_friends

Commit:
fix(Friend, BlockedMember) :

- blockedMember 경로 수정 (blocked > blockedmember)
- BlockedMemberExistDao 파일명 수정 (BlockedMemberExistDao > BlockedMemberValidateDao)
- FriendChangeStatusApi 변경할 상태 별로 클래스 분리
- predicateFunc 인자로 받는 메소드 수정

Commit:
fix(Friend) :

- 친구 추가 로직 리팩토링
- friend 변수명 수정

Commit:
fix(Friend) : 오타 수정

Commit:
fix(BlockedMember) : 유저 차단 해제 시 차단 여부 확인 로직  추가

holeman79 committed:
refactor(FriendValidateDao) :

- validateStatus 메소드 명을 좀 더 명시적으로 validateAcceptOrRejectFriend 변경
- validateAddFriend 조건문을 좀 더 간단하게 수정
- exception 메세지 수정

holeman79 committed:
refactor(BlockedMemberValidateDao) :

- 중복된 코드를 재활용할 수 있도록 메소드 추출
- exception 메세지 구분

holeman79 committed:
refactor(Friend Accept, Reject) :

- 당사자만 친구 수락 또는 거절 할 수 있도록 로직 추가

holeman79 committed:
refactor(FriendValidateDao) : FriendInputException 메세지에 memberId, friend 표기했던 것 제거

Commit:
refactor(BlockedMemberValidateDao) : 차단 해제 가능 여부 함수 및 테스트 코드 추가
